File: _code-samples/sponsored-fees-and-reserves/py/transfer_reserve_sponsorship.py
import json
import sys

from xrpl.clients import JsonRpcClient
from xrpl.models import (
    DepositPreauth,
    SponsorFlag,
    SponsorshipTransfer,
    SponsorshipTransferFlag,
)
from xrpl.transaction import autofill, sign, sign_as_sponsor, submit_and_wait
from xrpl.wallet import generate_faucet_wallet

# Set up client ----------------------
client = JsonRpcClient("https://s.devnet.rippletest.net:51234")

# Create the wallets ----------------------
print("\n=== Creating the sponsor and sponsee wallets... ===")
sponsor_a = generate_faucet_wallet(client)
sponsor_b = generate_faucet_wallet(client)
sponsee = generate_faucet_wallet(client)

print(f"Sponsor A address: {sponsor_a.address}")
print(f"Sponsor B address: {sponsor_b.address}")
print(f"Sponsee address:   {sponsee.address}")

# Create an unsponsored ledger entry ----------------------
# The sponsee creates a DepositPreauth entry with no sponsorship fields, so it pays
# the fee and the owner reserve for the resulting entry itself.
print("\n=== Submitting unsponsored DepositPreauth transaction... ===")
deposit_preauth_tx = DepositPreauth(
    account=sponsee.address,
    authorize=sponsor_a.address,
)
deposit_preauth_response = submit_and_wait(deposit_preauth_tx, client, sponsee)

if deposit_preauth_response.result["meta"]["TransactionResult"] != "tesSUCCESS":
    result_code = deposit_preauth_response.result["meta"]["TransactionResult"]
    print(f"Error: Unable to create the preauthorization: {result_code}")
    sys.exit(1)

preauth_node = next(
    node for node in deposit_preauth_response.result["meta"]["AffectedNodes"]
    if node.get("CreatedNode", {}).get("LedgerEntryType") == "DepositPreauth"
)
preauth_id = preauth_node["CreatedNode"]["LedgerIndex"]

print("DepositPreauth created successfully, with its reserve paid by the sponsee.")
print(f"DepositPreauth ID: {preauth_id}")

# Prepare SponsorshipTransfer transaction to start the sponsorship ----------------------
print("\n=== Preparing SponsorshipTransfer transaction to start the sponsorship... ===")
create_tx = SponsorshipTransfer(
    account=sponsee.address,
    object_id=preauth_id,
    flags=SponsorshipTransferFlag.TF_SPONSORSHIP_CREATE,
    sponsor=sponsor_a.address,
    sponsor_flags=SponsorFlag.SPF_SPONSOR_RESERVE,
)
create_tx = autofill(create_tx, client)

print(json.dumps(create_tx.to_xrpl(), indent=2))

# Sign as the sponsee, then co-sign as Sponsor A ----------------------
print("\n=== Submitting SponsorshipTransfer transaction... ===")
create_signed_tx = sign_as_sponsor(sponsor_a, sign(create_tx, sponsee))
create_response = submit_and_wait(create_signed_tx.tx, client)

if create_response.result["meta"]["TransactionResult"] != "tesSUCCESS":
    result_code = create_response.result["meta"]["TransactionResult"]
    print(f"Error: Unable to start the sponsorship: {result_code}")
    sys.exit(1)

fields = next(
    node["ModifiedNode"]["FinalFields"]
    for node in create_response.result["meta"]["AffectedNodes"]
    if node.get("ModifiedNode", {}).get("LedgerEntryType") == "DepositPreauth"
)
print("Sponsorship started successfully!")
print(f"DepositPreauth reserve now sponsored by: {fields['Sponsor']}")

# Prepare SponsorshipTransfer transaction to reassign the sponsorship ----------------
# tfSponsorshipReassign moves the reserve to Sponsor B in one transaction. Only the
# incoming sponsor has to consent; Sponsor A's obligation is released automatically.
print("\n=== Preparing SponsorshipTransfer transaction to reassign the sponsorship... ===")
reassign_tx = SponsorshipTransfer(
    account=sponsee.address,
    object_id=preauth_id,
    flags=SponsorshipTransferFlag.TF_SPONSORSHIP_REASSIGN,
    sponsor=sponsor_b.address,
    sponsor_flags=SponsorFlag.SPF_SPONSOR_RESERVE,
)
reassign_tx = autofill(reassign_tx, client)

print(json.dumps(reassign_tx.to_xrpl(), indent=2))

# Sign as the sponsee, then co-sign as Sponsor B ----------------------
print("\n=== Submitting SponsorshipTransfer transaction ===\n")
reassign_signed_tx = sign_as_sponsor(sponsor_b, sign(reassign_tx, sponsee))
reassign_response = submit_and_wait(reassign_signed_tx.tx, client)

if reassign_response.result["meta"]["TransactionResult"] != "tesSUCCESS":
    result_code = reassign_response.result["meta"]["TransactionResult"]
    print(f"Error: Unable to reassign the sponsorship: {result_code}")
    sys.exit(1)

fields = next(
    node["ModifiedNode"]["FinalFields"]
    for node in reassign_response.result["meta"]["AffectedNodes"]
    if node.get("ModifiedNode", {}).get("LedgerEntryType") == "DepositPreauth"
)
print("Sponsorship reassigned successfully!")
print(f"DepositPreauth reserve now sponsored by: {fields['Sponsor']}")

# Prepare SponsorshipTransfer transaction to end the sponsorship ----------------------
# tfSponsorshipEnd takes no Sponsor field and needs no co-signature. If it
# succeeds, the Sponsor field is removed and the sponsee becomes responsible for
# the entry's reserve again.
print("\n=== Preparing SponsorshipTransfer transaction to end the sponsorship... ===")
end_tx = SponsorshipTransfer(
    account=sponsee.address,
    object_id=preauth_id,
    flags=SponsorshipTransferFlag.TF_SPONSORSHIP_END,
)

print(json.dumps(end_tx.to_xrpl(), indent=2))

# Submit the SponsorshipTransfer transaction to end the sponsorship ----------------------
print("\n=== Submitting SponsorshipTransfer transaction... ===")
end_response = submit_and_wait(end_tx, client, sponsee)

if end_response.result["meta"]["TransactionResult"] != "tesSUCCESS":
    result_code = end_response.result["meta"]["TransactionResult"]
    print(f"Error: Unable to end the sponsorship: {result_code}")
    sys.exit(1)

fields = next(
    node["ModifiedNode"]["FinalFields"]
    for node in end_response.result["meta"]["AffectedNodes"]
    if node.get("ModifiedNode", {}).get("LedgerEntryType") == "DepositPreauth"
)
if "Sponsor" in fields:
    print("Error: The DepositPreauth entry still has a Sponsor field")
    sys.exit(1)

print("Sponsorship ended successfully!")
print("The sponsee now pays the DepositPreauth entry's owner reserve again.")

File: _code-samples/sponsored-fees-and-reserves/js/transferReserveSponsorship.js
import {
  Client,
  SponsorFlags,
  SponsorshipTransferFlags,
  signAsSponsor,
  validate
} from 'xrpl'

// Connect to the network ----------------------
const client = new Client('wss://s.devnet.rippletest.net:51233')
await client.connect()

// Create the wallets ----------------------
console.log(`\n=== Creating the sponsor and sponsee wallets... ===`)
const { wallet: sponsorA } = await client.fundWallet()
const { wallet: sponsorB } = await client.fundWallet()
const { wallet: sponsee } = await client.fundWallet()

console.log(`Sponsor A address: ${sponsorA.address}`)
console.log(`Sponsor B address: ${sponsorB.address}`)
console.log(`Sponsee address:   ${sponsee.address}`)

// Create an unsponsored ledger entry ----------------------
// The sponsee creates a DepositPreauth entry with no sponsorship fields, so it pays
// the fee and the owner reserve for the resulting entry itself.
console.log(`\n=== Submitting unsponsored DepositPreauth transaction... ===`)
const depositPreauthTx = {
  TransactionType: 'DepositPreauth',
  Account: sponsee.address,
  Authorize: sponsorA.address
}
const depositPreauthResponse = await client.submitAndWait(depositPreauthTx, {
  wallet: sponsee,
  autofill: true
})

if (depositPreauthResponse.result.meta.TransactionResult !== 'tesSUCCESS') {
  const resultCode = depositPreauthResponse.result.meta.TransactionResult
  console.error('Error: Unable to create the preauthorization:', resultCode)
  await client.disconnect()
  process.exit(1)
}

const preauthNode = depositPreauthResponse.result.meta.AffectedNodes.find(
  node => node.CreatedNode?.LedgerEntryType === 'DepositPreauth'
)
const preauthID = preauthNode.CreatedNode.LedgerIndex

console.log('DepositPreauth created successfully, with its reserve paid by the sponsee.')
console.log(`DepositPreauth ID: ${preauthID}`)
console.log(`Transaction URL: https://devnet.xrpl.org/transactions/${depositPreauthResponse.result.hash}`)

// Prepare SponsorshipTransfer transaction to start the sponsorship ----------------------
console.log(`\n=== Preparing SponsorshipTransfer transaction to start the sponsorship... ===`)
const createTx = {
  TransactionType: 'SponsorshipTransfer',
  Account: sponsee.address,
  ObjectID: preauthID,
  Flags: SponsorshipTransferFlags.tfSponsorshipCreate,
  Sponsor: sponsorA.address,
  SponsorFlags: SponsorFlags.tfSponsorReserve
}

validate(createTx)
const preparedCreateTx = await client.autofill(createTx)

console.log(JSON.stringify(preparedCreateTx, null, 2))

// Sign as the sponsee, then co-sign as Sponsor A ----------------------
console.log(`\n=== Submitting SponsorshipTransfer transaction... ===`)
const createSignedTx = signAsSponsor(sponsorA, sponsee.sign(preparedCreateTx).tx_blob)
const createResponse = await client.submitAndWait(createSignedTx.tx_blob)

if (createResponse.result.meta.TransactionResult !== 'tesSUCCESS') {
  const resultCode = createResponse.result.meta.TransactionResult
  console.error('Error: Unable to start the sponsorship:', resultCode)
  await client.disconnect()
  process.exit(1)
}

let fields = createResponse.result.meta.AffectedNodes.find(
  node => node.ModifiedNode?.LedgerEntryType === 'DepositPreauth'
).ModifiedNode.FinalFields
console.log('Sponsorship started successfully!')
console.log(`DepositPreauth reserve now sponsored by: ${fields.Sponsor}`)
console.log(`Transaction URL: https://devnet.xrpl.org/transactions/${createResponse.result.hash}`)

// Prepare SponsorshipTransfer transaction to reassign the sponsorship ----------------
// tfSponsorshipReassign moves the reserve to Sponsor B in one transaction. Only the
// incoming sponsor has to consent; Sponsor A's obligation is released automatically.
console.log(`\n=== Preparing SponsorshipTransfer transaction to reassign the sponsorship... ===`)
const reassignTx = {
  TransactionType: 'SponsorshipTransfer',
  Account: sponsee.address,
  ObjectID: preauthID,
  Flags: SponsorshipTransferFlags.tfSponsorshipReassign,
  Sponsor: sponsorB.address,
  SponsorFlags: SponsorFlags.tfSponsorReserve
}

validate(reassignTx)
const preparedReassignTx = await client.autofill(reassignTx)

console.log(JSON.stringify(preparedReassignTx, null, 2))

// Sign as the sponsee, then co-sign as Sponsor B ----------------------
console.log(`\n=== Submitting SponsorshipTransfer transaction... ===`)
const reassignSignedTx = signAsSponsor(sponsorB, sponsee.sign(preparedReassignTx).tx_blob)
const reassignResponse = await client.submitAndWait(reassignSignedTx.tx_blob)

if (reassignResponse.result.meta.TransactionResult !== 'tesSUCCESS') {
  const resultCode = reassignResponse.result.meta.TransactionResult
  console.error('Error: Unable to reassign the sponsorship:', resultCode)
  await client.disconnect()
  process.exit(1)
}

fields = reassignResponse.result.meta.AffectedNodes.find(
  node => node.ModifiedNode?.LedgerEntryType === 'DepositPreauth'
).ModifiedNode.FinalFields
console.log('Sponsorship reassigned successfully!')
console.log(`DepositPreauth reserve now sponsored by: ${fields.Sponsor}`)
console.log(`Transaction URL: https://devnet.xrpl.org/transactions/${reassignResponse.result.hash}`)

// Prepare SponsorshipTransfer transaction to end the sponsorship ----------------------
// tfSponsorshipEnd takes no Sponsor field and needs no co-signature. If it
// succeeds, the Sponsor field is removed and the sponsee becomes responsible for
// the entry's reserve again.
console.log(`\n=== Preparing SponsorshipTransfer transaction to end the sponsorship... ===`)
const endTx = {
  TransactionType: 'SponsorshipTransfer',
  Account: sponsee.address,
  ObjectID: preauthID,
  Flags: SponsorshipTransferFlags.tfSponsorshipEnd
}

validate(endTx)
console.log(JSON.stringify(endTx, null, 2))

// Submit the SponsorshipTransfer transaction to end the sponsorship ----------------------
console.log(`\n=== Submitting SponsorshipTransfer transaction... ===`)
const endResponse = await client.submitAndWait(endTx, {
  wallet: sponsee,
  autofill: true
})

if (endResponse.result.meta.TransactionResult !== 'tesSUCCESS') {
  const resultCode = endResponse.result.meta.TransactionResult
  console.error('Error: Unable to end the sponsorship:', resultCode)
  await client.disconnect()
  process.exit(1)
}

fields = endResponse.result.meta.AffectedNodes.find(
  node => node.ModifiedNode?.LedgerEntryType === 'DepositPreauth'
).ModifiedNode.FinalFields
if (fields.Sponsor !== undefined) {
  console.error('Error: The DepositPreauth entry still has a Sponsor field')
  await client.disconnect()
  process.exit(1)
}

console.log('Sponsorship ended successfully!')
console.log(`The sponsee now pays the DepositPreauth entry's owner reserve again.`)
console.log(`Transaction URL: https://devnet.xrpl.org/transactions/${endResponse.result.hash}`)

await client.disconnect()
